### ISSUE TITLE
Provide functions for computing costs based on the edges taken

### ISSUE THREAD

**Issue by isovector** [devonhollowood/search-algorithms#13]
Right now costs are given by a function `state -> state -> cost`; in my case, I know the cost of an edge taken, but computing the cost given two states is prohibitively expensive.

**Comment by isovector** [comment]
I've come up with the following:

```haskell
edgeDijkstra
    :: (Foldable f, Num cost, Ord cost, Ord state)
    => (state -> f (cost, state))
    -> (state -> Bool)
    -> state
    -> Maybe (cost, [state])
edgeDijkstra e s i
  = fmap (second $ fmap snd)
  $ dijkstra (e . snd) (const fst) (s . snd) (0, i)
```

**Comment by isovector** [comment]
and while i'm at it, here's a bounded dfs:

```haskell
boundedDfs
  :: (Foldable f, Ord state)
  => Int  -- ^ max depth
  -> (state -> f state)
  -> (state -> Bool)
  -> state
  -> Maybe [state]
boundedDfs d e f i = fmap (fmap snd) $
  dfs
    (\(ix, s) ->
      case ix >= d of
        True  -> []
        False -> fmap (ix + 1, ) $ toList $ e s)
    (f . snd)
    (0 :: Int, i)
```

**Comment by devonhollowood** [comment]
All the above seems reasonable. Thanks! Do you want to write a pull request containing the above + some documentation, so that you get contribution credit? Otherwise, I can do it.

**Comment by isovector** [comment]
I'm not going to get a chance to tackle this for about a week; I'm ok if you want to merge it sooner. Thanks!

**Comment by devonhollowood** [comment]
Alright, in that case, I'll do it if I get a chance this week; otherwise I'll just wait for you to do it.

**Comment by devonhollowood** [comment]
Fixed via #14
